File: smarttrade_app/lib/features/trading_journal/data/datasources/asset_preference_local_datasource.dart
import 'package:sqflite/sqflite.dart';
import 'package:smarttrade_app/core/database/app_database.dart';
import 'package:smarttrade_app/features/trading_journal/data/models/asset_preference_model.dart';
import 'package:smarttrade_app/features/trading_journal/domain/entities/asset_preference.dart';

class AssetPreferenceLocalDataSource {
  const AssetPreferenceLocalDataSource(this._database);

  final Database _database;

  Future<List<AssetPreference>> getAll() async {
    final rows = await _database.query(
      AppDatabaseTables.assetPreferences,
      orderBy: 'display_order ASC',
    );
    return rows.map(AssetPreferenceModel.fromMap).toList();
  }

  Future<void> save(AssetPreference preference) async {
    await _database.insert(
      AppDatabaseTables.assetPreferences,
      AssetPreferenceModel.toMap(preference),
      conflictAlgorithm: ConflictAlgorithm.replace,
    );
  }
}

File: smarttrade-web/src/features/trading-journal/data/datasources/assetPreferenceLocalDataSource.ts
import { db } from '@/core/persistence/db'
import type { AssetPreference } from '../../domain/entities/assetPreference'

export const assetPreferenceLocalDataSource = {
  async getAll(): Promise<AssetPreference[]> {
    return db.assetPreferences.orderBy('displayOrder').toArray()
  },

  async save(preference: AssetPreference): Promise<void> {
    await db.assetPreferences.put(preference)
  },
}
